### ISSUE TITLE
Can't go from a child to a parent's sibling which redirects to a sibling of the child

### ISSUE THREAD

**Issue by stephenbunch** [angular-ui/ui-router#1234]
I encountered what seems to be a bug where if I navigate to a child, and then navigate to a sibling of the parent which redirects back to a sibling of the child, nothing is rendered (`#3`). At which point, I must navigate to a different sibling of the parent before being able to render any of the children again (`#9`).

```js
$stateProvider.state( 'a', {
  template: '<ui-view/>'
});

$stateProvider.state( 'a.b', {
  template: '<h1>hello world</h1>'
});

$stateProvider.state( 'a.c', {
  template: '<h1>bla bla bla</h1>'
});

$stateProvider.state( 'd', {
  onEnter: [ '$state', function( $state ) {
    $state.go( 'a.c' );
  }]
});

$stateProvider.state( 'e', {
  template: '<h1>:)</h1>'
});

$stateProvider.state( 'a.f', {
  onEnter: [ '$state', function( $state ) {
    $state.go( 'a.c' );
  }]
});

// 1. a.c -> works
// 2. a.b -> works
// 3. a.b -> d -> doesn't work
// 4. e -> d -> works
// 5. a.b -> a.f -> works
// 6. a.b -> d -> a -> TypeError: Cannot read property '@' of null
// 7. a.b -> d -> a.f -> still doesn't work
// 8. a.b -> d -> d, {reload: true} -> a.f -> nope
// 9. a.b -> d -> e -> a.c -> yay, it renders again!
```

**Comment by oliviercpl** [comment]
Same bug , how to solve this !

**Comment by laravelfanatic** [comment]
@oliviercpl
Using Controller seems to work :)

```javascript
.state("redirect",{
  template: "whatever",
  params: { 'targetState':true, 'targetParams': true},
  controller: ['$state', '$stateParams', function($state, $stateParams){
    $state.go($stateParams.targetState, $stateParams.targetParams);
  }],
})
```

**Comment by majo44** [comment]
Using Controller works, but is going through the redirect step, so the template is loaded, controllers inside are initialized etc. The cost is huge. Any update, I still have this problem :(

**Comment by christopherthielen** [comment]
If you don't need resolves to know whether to redirect, you can redirect using a $stateChangeStart event listener.

**Comment by majo44** [comment]
I checked this solution, but the problem is that if:
- I will call preventDefault  the url is not changing (eg by click on ui-sref). 
- If I will not use preventDefault, my call on $state.go in handler have no any effect.
- If I will not use preventDefault and call $state.go in handler in the $timeout(0) the url is not changing.

I have the scenario where on attempt to go to some state (url is chaging) I checking authorization, if user is not authorized, I want to go to some intermediate state (state without url) eg. for showing login screen, but without url changing, after authorize I simply calling $urlRouter.sync() and everything works. Similar scenario I have with session expired. Etc... I do not want to care about url's by self, and store them somehow as search parameter :)

Maybe I'm using bad approach for solving my scenario.

Btw. @christopherthielen ui-router is very usefull ang good solution, thx :)

**Comment by majo44** [comment]
I found the solution, maybe ugly, but works:

Solution:
```javascript

appmodule.config(function ($urlRouterProvider) {
    $urlRouterProvider.deferIntercept();
});

appmodule.run(
    function ($rootScope, $state, $injector, $urlRouter, $location) {

        var blockNextUrlUpdate;
        $rootScope.$on('$locationChangeSuccess', function (e) {
            if (blockNextUrlUpdate) {
                e.preventDefault();
                blockNextUrlUpdate = false;
            }
        });
        $urlRouter.listen();

        $rootScope.$on("$stateChangeStart", function (
            event, toState, toParams, fromState, fromParams)  {

            var wasPrevented = event.defaultPrevented;

            function doOnBefore(state) {
                var parent = state.parent,
                    name = state.name;

                parent =
                    (angular.isString(parent)) ? $state.get(parent)
                        : (angular.isObject(parent)) ? parent
                        : (name.indexOf('.') !== -1) ? $state.get(
                              name.substring(0, name.lastIndexOf('.')))
                        : null;

                if (parent) {
                    doOnBefore(parent);
                }

                if (!event.defaultPrevented) {
                    if (state["onBefore"]) {
                        $injector.invoke(state["onBefore"], null, {
                            event: event,
                            toState: toState,
                            toParams: toParams,
                            fromState: fromState,
                            fromParams: fromParams
                        });
                    }
                }
            }

            doOnBefore(toState);

            if (!wasPrevented && event.defaultPrevented) {
                url = $state.href(toState, toParams);
                    $location.url(url.substr(1));
                    blockNextUrlUpdate = true;
            }
        });
    });
```

Updated plunker from @stephenbunch : http://plnkr.co/edit/x9VyBGsxVz4lbU5WDKdW?p=preview

**Comment by majo44** [comment]
Previous my solution is not working well with browser back, forward :(

**Comment by christopherthielen** [comment]
UI-Router 1.0 preview is available: https://github.com/angular-ui/ui-router/tree/feature-1.0

**Comment by nateabele** [comment]
This should be resolved in 1.0. Unfortunately the underlying flaw in 0.2 is too problematic to address with our limited resources. Please feel free to post the workaround to the FAQ for 0.2 users. Sorry for your trouble.
